### PR TITLE
Create api hooks for react query

### PR DESCRIPTION
**Pull request by Nepomuceno**
### What does this PR do?

In this pull request, I have made changes to the project by adding queries in the form of React Query. These queries eliminate the need to use Redux in many areas of the project and provide caching functionality. The main advantage of these changes is that we no longer have a single record reprersented in the entire applicaion. Instead, each record is divided into different its own queries based on the file name, account, and container being accessed.

By using React Query, we can now prefetch information and avoid redundant metadata refetching. Previously, metadata had to be fetched twice: once to display in the list of files and again when navigating to the spectrogram page. With the new changes, we can enable direct links to the spectrogram for any of the built-in data sources.

These modifications improve the efficiency and performance of the application by leveraging the capabilities of React Query, reducing reliance on Redux, and providing a more granular approach to handling queries and caching.

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you successfully ran tests with your changes locally? Including container validation.
* [x] Have you lint your code locally prior to submission?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you labelled this PR correctly? E.G. major, milestone, breaking, breaking-change, minor, feature, enhancement, patch, fix, chore, refactor etc

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

### CODE REVIEW

File: client/src/Utils/sigmfMetadata.ts
Comment (Jaydee123) [inline client/src/Utils/sigmfMetadata.ts:1]: where's the tests? ;)

File: client/src/Utils/sigmfMetadata.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    zoomLevel: number
+  ) {
+    let result = {
+      x1: 0,
+      x2: 0,
+      y1: 0,
+      y2: 0,
+      visible: false,
+    };
+    let freq_lower_edge = this['core:freq_lower_edge'];
```
Comment (Jaydee123) [inline client/src/Utils/sigmfMetadata.ts:204]: how do we gracefully handle if these things aren't set?

File: client/src/Utils/sigmfMetadata.ts
Comment (Nepomuceno) [reply client/src/Utils/sigmfMetadata.ts:1]: THey are all passing now ;)

File: client/src/Utils/sigmfMetadata.ts (hunk 1 above)
Comment (Nepomuceno) [reply client/src/Utils/sigmfMetadata.ts:204]: those fields are required for annotations. you cannot create an annotation object without them

File: client/src/Utils/sigmfMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+export class TraceabilityOrigin {
+  type: string;
+  account: string;
+  container: string;
+  file_path: string;
+}
+
+export class Annotation {
```
Comment (Jaydee123) [inline client/src/Utils/sigmfMetadata.ts:177]: Is it worth exporting this into it's own file? this file's a bit of a chunky boi

File: client/src/api/datasource/LocalClient.ts [hunk 3]
```diff
@@ -1,22 +1,99 @@
...
+    if (!metadataFile) {
+      return Promise.reject('No file found');
+    }
+    const dataFile = localDirectory.find((file) => {
+      return file.webkitRelativePath === filePath + '.sigmf-data';
+    });
+    if (!dataFile) {
+      return Promise.reject('No file found');
+    }
+    return this.getMetaFromFile(metadataFile, dataFile, account, container);
```
Comment (Jaydee123) [inline client/src/api/datasource/LocalClient.ts:94]: I do not understand why the linter doesn't complain here, MAKE IT MAKE SENSE

File: client/src/api/iqdata/BlobClient.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    // if filePath does not finish in .sigmf-data, add it
+    if (!file_path.endsWith('.sigmf-data')) {
+      file_path += '.sigmf-data';
+    }
+    let startTime = performance.now();
+    const bytesPerSample = meta.getBytesPerSample();
+    const offsetBytes = index * tileSize * bytesPerSample * 2;
+    const countBytes = tileSize * bytesPerSample * 2;
+    const blobClient = getBlobClient(account, container, file_path);
+    // Thi is ugly but it is the only way to get the blob as an ArrayBuffer
```
Comment (Jaydee123) [inline client/src/api/iqdata/BlobClient.ts:24]: that's not very nice

File: client/src/api/datasource/LocalClient.ts [hunk 5]
```diff
@@ -1,22 +1,99 @@
...
+    file: FileWithDirectoryAndFileHandle,
+    dataFile: FileWithDirectoryAndFileHandle,
+    account: string,
+    container: string
+  ) {
+    const fileContent = await file.text();
+    let metadata = Object.assign(new SigMFMetadata(), JSON.parse(fileContent)) as SigMFMetadata;
+    metadata.dataFileHandle = dataFile;
+    metadata.metadataFileHandle = file;
+    const origin: TraceabilityOrigin = metadata['traceability:origin'];
```
Comment (Jaydee123) [inline client/src/api/datasource/LocalClient.ts:18]: this could be null

File: client/src/api/datasource/LocalClient.ts (hunk 3 above)
Comment (Nepomuceno) [reply client/src/api/datasource/LocalClient.ts:94]: Yeha I thought aboput that the file kept growing btu changint i now would reuire to change the iomports of the changes that were done already in the other commit so I would say this si something that could be done after 👍

File: client/src/api/iqdata/BlobClient.ts (hunk 4 above)
Comment (Nepomuceno) [reply client/src/api/iqdata/BlobClient.ts:24]: I knwo blame the people that wrote the auzre blob API

File: client/src/Utils/sigmfMetadata.ts
Comment (Nepomuceno) [reply client/src/Utils/sigmfMetadata.ts:1]: Added tests :)

File: client/src/api/datasource/LocalClient.ts (hunk 5 above)
Comment (Nepomuceno) [reply client/src/api/datasource/LocalClient.ts:18]: I check that in the line bellow

File: client/src/Utils/sigmfMetadata.ts (hunk 2 above)
Comment (Nepomuceno) [reply client/src/Utils/sigmfMetadata.ts:177]: Yeha I thought about that the file kept growing but changin it now would require to change the imports of the changes that were done already in the other commit so I would say this is something that could be done after

File: client/src/api/datasource/LocalClient.ts (hunk 3 above)
Comment (Nepomuceno) [reply client/src/api/datasource/LocalClient.ts:94]: It should not complain here I am glad ity doesnt we need to chack in the other times when it does complain for an else that is not neded.

**Review by Jaydee123 — APPROVED** [review]

**Pull request merged**